File: wicket-extensions/src/test/java/org/apache/wicket/extensions/markup/html/form/palette/PaletteTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.wicket.extensions.markup.html.form.palette;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import org.apache.wicket.WicketTestCase;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.LoadableDetachableModel;
import org.apache.wicket.model.util.ListModel;
import org.apache.wicket.util.tester.FormTester;
import org.apache.wicket.validation.IValidatable;
import org.apache.wicket.validation.IValidator;
import org.apache.wicket.validation.ValidationError;
import org.junit.Test;

/**
 * Test for {@link Palette}.
 */
public class PaletteTest extends WicketTestCase
{

	/**
	 */
	@Test
	public void standard()
	{
		IModel<List<String>> selected = new ListModel<>(new ArrayList<>(Arrays.asList("A", "D")));

		IModel<List<String>> all = new ListModel<>(new ArrayList<>(
			Arrays.asList("A", "B", "C", "D")));

		PaletteTestPage testPage = new PaletteTestPage(selected, all);

		tester.startPage(testPage);

		FormTester formTester = tester.newFormTester(testPage.form.getId());
		formTester.submit();

		Collection<String> collection = testPage.palette.getModelCollection();

		assertEquals(2, collection.size());
		Iterator<String> iterator = collection.iterator();
		assertEquals("A", iterator.next());
		assertEquals("D", iterator.next());
	}

	/**
	 * WICKET-4231 palette with choicesModel missing currently selected
	 */
	@Test
	public void choicesModelMissingSelected()
	{
		IModel<List<String>> selected = new ListModel<>(new ArrayList<>(Arrays.asList("D")));

		IModel<List<String>> all = new ListModel<>(new ArrayList<>(Arrays.asList("A", "B", "C")));

		PaletteTestPage testPage = new PaletteTestPage(selected, all);

		tester.startPage(testPage);

		FormTester formTester = tester.newFormTester(testPage.form.getId());
		formTester.submit();

		Collection<String> collection = testPage.palette.getModelCollection();

		assertEquals(0, collection.size());
	}

	/**
	 * WICKET-4231 palette with choicesModel accounting for currently selected
	 */
	@Test
	public void choicesModelAccountingForSelected()
	{
		final List<String> list = new ArrayList<>(Arrays.asList("D"));

		IModel<List<String>> selected = new ListModel<>(list);

		IModel<List<String>> all = new LoadableDetachableModel<List<String>>()
		{
			@Override
			protected List<String> load()
			{
				List<String> fromDB = Arrays.asList("A", "B", "C"); // normally coming from DB

				List<String> result = new ArrayList<>();
				result.addAll(fromDB);

				// include already selected
				result.addAll(list);

				return result;
			}
		};

		PaletteTestPage testPage = new PaletteTestPage(selected, all);

		tester.startPage(testPage);

		FormTester formTester = tester.newFormTester(testPage.form.getId());
		formTester.submit();

		Collection<String> collection = testPage.palette.getModelCollection();

		assertEquals(1, collection.size());
		assertEquals("D", collection.iterator().next());
	}

	/**
	 * WICKET-4590 single unselected item
	 */
	@Test
	public void choicesModelSingleNotSelected()
	{
		IModel<List<String>> selected = new ListModel<>(new ArrayList<String>());

		IModel<List<String>> all = new ListModel<>(new ArrayList<>(Arrays.asList("A")));

		PaletteTestPage testPage = new PaletteTestPage(selected, all);

		tester.startPage(testPage);

		tester.assertContains("<option value=\"A\">A</option>");
	}

	@Test
	public void required()
	{
		IModel<List<String>> selected = new ListModel<>(new ArrayList<String>());

		IModel<List<String>> all = new ListModel<>(new ArrayList<>(Arrays.asList("A")));

		PaletteTestPage testPage = new PaletteTestPage(selected, all);
		testPage.palette.setRequired(true);

		tester.startPage(testPage);

		FormTester formTester = tester.newFormTester(testPage.form.getId());
		formTester.submit();

		assertTrue(testPage.form.hasError());
	}

	@Test
	public void validationErrorRawInput()
	{
		IModel<List<String>> selected = new ListModel<>(new ArrayList<String>());

		IModel<List<String>> all = new ListModel<>(new ArrayList<>(Arrays.asList("A", "B")));

		PaletteTestPage testPage = new PaletteTestPage(selected, all);
		testPage.palette.add(new IValidator<Collection<String>>()
		{
			@Override
			public void validate(IValidatable<Collection<String>> validatable)
			{
				if (validatable.getValue().contains("A"))
				{
					validatable.error(new ValidationError("A not allowed"));
				}
			}
		});

		tester.startPage(testPage);

		FormTester formTester = tester.newFormTester(testPage.form.getId());
		formTester.setValue("palette:recorder", "A");
		formTester.submit();

		assertTrue(testPage.form.hasError());

		// with RAW_INPUT
		tester.assertContains("<option value=\"B\">B</option>\\s*</select>");
		tester.assertContains("<option value=\"A\">A</option>\\s*</select>");

		testPage.form.clearInput();

		tester.startPage(testPage);

		// without RAW_INPUT
		tester
			.assertContains("<option value=\"A\">A</option>\\s*<option value=\"B\">B</option>\\s*</select>");
	}
}

File: wicket-extensions/src/test/js/palette-test.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*global QUnit: true */

jQuery(document).ready(function() {
	"use strict";

	const { module, test } = QUnit;

	function selectionOptionValues() {
		var selection = document.getElementById('paletteSelection');
		return Array.prototype.map.call(selection.options, function(option) {
			return option.value;
		});
	}

	function selectOnly(selection, value) {
		Array.prototype.forEach.call(selection.options, function(option) {
			option.selected = (option.value === value);
		});
	}

	module("Wicket.Palette", {
		beforeEach: function() {
			var selection = document.getElementById('paletteSelection');
			selection.innerHTML =
				'<option value="1">one</option>' +
				'<option value="2">two</option>' +
				'<option value="3">three</option>';
			document.getElementById('paletteRecorder').value = '1,2,3';
		}
	});

	test("moveUp reorders the selected option without throwing", assert => {
		var selection = document.getElementById('paletteSelection');
		selectOnly(selection, '3');

		Wicket.Palette.moveUp('paletteChoices', 'paletteSelection', 'paletteRecorder');

		assert.deepEqual(selectionOptionValues(), ['1', '3', '2'],
			"moveUp did not move the selected option in front of its predecessor");
	});

	test("moveUp updates the hidden recorder input", assert => {
		var selection = document.getElementById('paletteSelection');
		selectOnly(selection, '3');

		Wicket.Palette.moveUp('paletteChoices', 'paletteSelection', 'paletteRecorder');

		assert.equal(document.getElementById('paletteRecorder').value, '1,3,2',
			"the recorder was not updated, so the server never sees the new order");
	});

	test("moveUp on the first option is a no-op and does not throw", assert => {
		var selection = document.getElementById('paletteSelection');
		selectOnly(selection, '1');

		Wicket.Palette.moveUp('paletteChoices', 'paletteSelection', 'paletteRecorder');

		assert.deepEqual(selectionOptionValues(), ['1', '2', '3'],
			"the order should not have changed");
	});

	test("moveDown reorders the selected option without throwing", assert => {
		var selection = document.getElementById('paletteSelection');
		selectOnly(selection, '1');

		Wicket.Palette.moveDown('paletteChoices', 'paletteSelection', 'paletteRecorder');

		assert.deepEqual(selectionOptionValues(), ['2', '1', '3'],
			"moveDown did not move the selected option behind its successor");
	});
});
